### PR TITLE
Test extraction

### PR DESCRIPTION
**Pull request by oysteoh**

### CODE REVIEW

File: tests/storage/test_extraction.py [hunk 1]
```diff
@@ -201,7 +247,105 @@ def test_parameters(client):
...
+    # Validate data
+    obs = observations[0]
+    assert obs["name"] == "OBS"
+    assert obs["values"] == data[0].tolist()
+    assert obs["errors"] == data[1].tolist()
+    assert obs["x_axis"] == data.index.astype(str).tolist()
+    assert obs["transformation"] is None
+
+
+def test_update(client):
```
Comment (xjules) [inline tests/storage/test_extraction.py:277]: Rename `test_post_update_data`

File: tests/storage/test_extraction.py [hunk 2]
```diff
@@ -201,7 +255,140 @@ def test_parameters(client):
...
+    # Parent ensemble should have a child
+    assert ensembles[0]["child_ensemble_ids"] == [child_ensemble_id]
+    assert ensembles[0]["parent_ensemble_id"] is None
+
+    # Child ensemble should have a parent
+    assert ensembles[1]["child_ensemble_ids"] == []
+    assert ensembles[1]["parent_ensemble_id"] == parent_ensemble_id
+
+
+def test_observation_transformation(client):
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:348]: Remove this test. It's a WIP and doesn't belong in this PR.

File: tests/storage/test_extraction.py [hunk 3]
```diff
@@ -40,18 +42,35 @@ class ErtConfigBuilder:
...
         self._priors[name] = entry
         return self
 
-    def build(self, path=None):
+    def build(self, path=None, path_executable=None):
         if path is None:
             path = Path.cwd()
-        (path / "JOB").write_text("EXECUTABLE /usr/bin/true\n")
+
+        if path_executable is None:
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:67]: Job generating code can be moved to `_build_job`, which:
1. Creates a file called `JOB`, pointing to `/usr/bin/true` by default, or to the locally created executable if `self._job_script` is set. The `self._job_script` lets us specify the code for a job to be run.
2. Appends "INSTALL_JOB" and "SIMULATION_JOB" to the ERT config file.

File: tests/storage/test_extraction.py [hunk 4]
```diff
@@ -296,7 +483,7 @@ def _rand_name():
     return "".join(random.choice(string.ascii_lowercase) for _ in range(8))
 
 
-def _create_runpath(ert: LibresFacade):
+def _create_runpath(ert: LibresFacade, iteration: int = 0, run: bool = False):
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:486]: `_create_runpath` should just create the runpath. Rename it for the `run` to make sense, or create a separate function called `_evaluate_ensemble`.

**Review by dotfloat — CHANGES_REQUESTED** [review]

File: tests/storage/test_extraction.py [hunk 5]
```diff
@@ -201,7 +261,104 @@ def test_parameters(client):
         df = pd.read_csv(stream, index_col=0, float_precision="round_trip")
 
         # ERT produces a low-quality version
-        assert_almost_equal(parameters_df[col].values, df.values.flatten(), decimal=4)
+        assert_almost_equal(parameters_df[col].values, df.values.flatten(), decimal=5)
+
+
+def test_observations(client):
```
Comment (xjules) [inline tests/storage/test_extraction.py:267]: I would also rename this test function to give it more semantics, eg. `test_post_observations`?

File: tests/storage/test_extraction.py [hunk 6]
```diff
@@ -201,7 +261,104 @@ def test_parameters(client):
...
+        f"""\
+    #!/usr/bin/python3
+
+    if __name__ == "__main__":
+        output = {str(data)}
+        with open("poly_0.out", "w") as f:
+            f.write("\\n".join(map(str, output)))
+    """
+    )
+    builder._job_script = script
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:314]: Change `_job_script` to `job_script`. Ie, unprivatise it.

File: tests/storage/test_extraction.py [hunk 7]
```diff
@@ -201,7 +261,104 @@ def test_parameters(client):
...
+    assert obs["transformation"] is None
+
+
+def test_post_ensemble_results(client):
+    data = pd.DataFrame([[1, 0.1], [2, 0.2], [3, 0.3], [4, 0.4]], index=[2, 4, 6, 8])
+
+    # Add priors to ERT config
+    builder = ErtConfigBuilder()
+    builder.ensemble_size = 2
+    response_name = "RES"
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:297]: Style: Move `response_name` to before the `builder` maybe? I personally think it's nicer if we don't mix assigning to constants with other code.

Feel free to ignore. 👍

File: tests/storage/test_extraction.py [hunk 8]
```diff
@@ -70,11 +87,54 @@ def _build_ert(self, path):
...
 
+    def _build_job(self, path):
+        f = (path / "test.ert").open("a")
+        f.write("INSTALL_JOB job JOB\n" "SIMULATION_JOB job\n")
+
+        if self._job_script is None:
+            (path / "JOB").write_text("EXECUTABLE /usr/bin/true\n")
+        else:
+            (path / "JOB").write_text(f"EXECUTABLE {path}/script\n")
+            (path / "script").write_text(self._job_script)
```
Comment (dotfloat) [inline tests/storage/test_extraction.py:103]: I think you need `(path / "script").chmod(0o755)`? Or is ERT somehow able to execute it without permissions?

File: ert_shared/models/base_run_model.py [hunk 9]
```diff
@@ -1,4 +1,6 @@
 from typing import Optional
+
+from ert3 import algorithms
```
Comment (dotfloat) [inline ert_shared/models/base_run_model.py:3]: Unused? Probably shouldn't import from `ert3` in `ert_shared`.

**Review by dotfloat — APPROVED** [review]
👑

**Pull request merged**